pkgs/desktops/gnome-3/misc/gnome-screensaver/gnome-xscreensaver/gnomexscreensaver.py

```diff
@@ -0,0 +1,229 @@
+import subprocess
+from datetime import datetime
+import threading
+from time import sleep
+import os
+import dbus
+import dbus.service
+from dbus.mainloop.glib import DBusGMainLoop
+from gi.repository import GLib
+
+state_lock = threading.Lock()
+blank_time = None
+
+inhibition_lock = threading.Lock()
+inhibition_cookie = 0
+inhibitions = {}
+
+xscreensaver_bin = '@XSCREENSAVER_BIN@'
+if xscreensaver_bin == '@' + 'XSCREENSAVER_BIN' + '@':
+  xscreensaver_bin = ''
+
+def command_args(*args):
+  return [ xscreensaver_bin + "xscreensaver-command" ] + list(args)
+
+def command(*args):
+  print("command:", args)
+  subprocess.check_call(command_args(*args))
+
+def Lock():
+  command('-lock')
+
+def Cycle():
+  command('-cycle')
+
+def SimulateUserActivity():
+  command('-deactivate')
+
+def Inhibit(app_name, reason):
+  command('-deactivate')
+  with inhibition_lock:
+    cookie = inhibition_cookie
+    inhibition_cookie += 1
+    inhibitions[cookie] = ( app_name, reason )
+    return cookie
+
+def UnInhibit(cookie):
+  with inhibition_lock:
+    inhibitions.pop(cookie, None)
+
+def Throttle(app_name, reason, cookie):
+  # TODO, not sure if this is possible
+  pass
+
+def UnThrottle(cookie):
+  # TODO, not sure if this is possible
+  pass
+
+def SetActive(activate):
+  if activate:
+    command('-activate')
+  else:
+    command('-deactivate')
+
+def GetActive():
+  with state_lock:
+    return blank_time != None
+
+def GetActiveTime():
+  with state_lock:
+    if blank_time == None:
+      return 0
+    else:
+      time = datetime.now()
+      return int((time - blank_time).total_seconds())
+
+def GetSessionIdle():
+  # TODO: is this possible to implement?
```

Comment:
Maybe using `-watch`?

Reply:
Actually `-time` prints details `XScreenSaver 5.43: screen non-blanked since Sat Dec 28 08:22:48 2019 (hack #186)`

Reply:
It doesn't report idleness unfortunately.

Reply:
I would say `XScreenSaver 5.43: screen blanked since Sat Dec 28 08:26:18 2019 (hack #211)` is idle.

I would expect idle to mean blank, since there is no other similar thing in the dbus api: https://gitlab.gnome.org/Archive/gnome-screensaver/blob/master/doc/dbus-interface.xml

Reply:
I think idle means the user not moving (even before the screensaver turns on). What you're talking about is `GetActiveTime`. I'm currently looking at getting the idle time either with `xprintidle` or with dbus from the gnome session manager.